File: windows/tests/Glyphstroke.Core.Tests/UpdateCheckTests.cs
using Glyphstroke.Core;
using Xunit;

namespace Glyphstroke.Core.Tests;

/// <summary>Проверка обновлений: сравнение версий, разбор ответа, перерыв.</summary>
public class UpdateCheckTests : IDisposable
{
    private readonly string _root = Path.Combine(Path.GetTempPath(), "glyphstroke-update-" + Guid.NewGuid());

    public void Dispose()
    {
        if (Directory.Exists(_root))
        {
            Directory.Delete(_root, recursive: true);
        }
    }

    [Fact]
    public void VersionParsing()
    {
        Assert.Equal(new[] { 0, 7, 0 }, UpdateCheck.Parse("0.7.0"));
        Assert.Equal(new[] { 1, 2, 3 }, UpdateCheck.Parse("v1.2.3"));
        Assert.Equal(new[] { 0 }, UpdateCheck.Parse(""));
    }

    [Fact]
    public void PrereleaseIsNotNewerThanRelease()
    {
        // «0.8.0-rc1» не должен выглядеть новее готовой 0.8.0.
        Assert.False(UpdateCheck.IsNewer("0.8.0-rc1", "0.8.0"));
        Assert.True(UpdateCheck.IsNewer("0.8.0", "0.7.9"));
        Assert.False(UpdateCheck.IsNewer("0.7.0", "0.7.0"));
        Assert.True(UpdateCheck.IsNewer("1.0", "0.9.9"));
        Assert.False(UpdateCheck.IsNewer("0.9", "0.9.1"));
    }

    [Fact]
    public void ReleaseIsReadFromTheAnswer()
    {
        var release = UpdateCheck.ReadRelease(
            """{"tag_name": "v0.9.0", "html_url": "https://пример/0.9.0", "body": " что нового "}""");
        Assert.Equal("0.9.0", release?.Version);
        Assert.Equal("https://пример/0.9.0", release?.Url);
        Assert.Equal("что нового", release?.Notes);
    }

    [Fact]
    public void AnswerWithoutATagIsIgnored()
    {
        Assert.Null(UpdateCheck.ReadRelease("""{"message": "Not Found"}"""));
        Assert.Null(UpdateCheck.ReadRelease("не json"));
    }

    [Fact]
    public void VersionComesFromTheAssetWhenAsked()
    {
        // Общий тег на три системы (v0.3.2) не должен выдавать себя за версию
        // винды: её берём из имени установщика в релизе.
        const string json = """
        {"tag_name": "v0.3.2", "html_url": "u", "body": "",
         "assets": [{"name": "glyphstroke_0.8.10_all.deb"},
                    {"name": "Glyphstroke-0.3.0.0-setup.exe"}]}
        """;
        Assert.Equal("0.3.0.0", UpdateCheck.ReadRelease(json, "-setup.exe")?.Version);
        // Без маркера — прежнее поведение, версия из тега.
        Assert.Equal("0.3.2", UpdateCheck.ReadRelease(json)?.Version);
        // Установленная 0.3.0 и .exe 0.3.0.0 — обновления нет.
        Assert.False(UpdateCheck.IsNewer("0.3.0.0", "0.3.0"));
    }

    [Fact]
    public void IntervalIsRespected()
    {
        var store = new Store(_root);
        Assert.True(UpdateCheck.Due(store), "ни разу не спрашивали");

        UpdateCheck.Remember(store, null);
        Assert.False(UpdateCheck.Due(store));
        Assert.True(UpdateCheck.Due(store, DateTimeOffset.UtcNow.Add(UpdateCheck.Interval).AddMinutes(1)));
        // время перевели назад: лучше спросить, чем ждать сутки от будущего
        Assert.True(UpdateCheck.Due(store, DateTimeOffset.UtcNow.AddHours(-5)));
    }

    [Fact]
    public void PendingKeepsOnlyNewerVersions()
    {
        var store = new Store(_root);
        UpdateCheck.Remember(store, new ReleaseInfo("9.9.9", "https://пример/9.9.9", string.Empty));

        Assert.Equal("9.9.9", UpdateCheck.Pending(store, "0.1.0")?.Version);
        Assert.Null(UpdateCheck.Pending(store, "9.9.9"));     // своя же версия — не новость
        Assert.Null(UpdateCheck.Pending(store, "10.0.0"));
    }

    [Fact]
    public void RepositoryComesFromSettings()
    {
        var settings = new Settings();
        Assert.Equal(UpdateCheck.DefaultRepository, UpdateCheck.Repository(settings));
        settings.UpdateRepo = " кто-то/что-то ";
        Assert.Equal("кто-то/что-то", UpdateCheck.Repository(settings));
        Assert.EndsWith("/releases/latest", UpdateCheck.ReleasePage("a/b"));
        Assert.StartsWith("https://api.github.com/repos/", UpdateCheck.ApiUrl("a/b"));
    }
}

File: macos/Tests/GlyphstrokeCoreTests/UpdateCheckTests.swift
import XCTest
@testable import GlyphstrokeCore

/// Проверка обновлений: сравнение версий, разбор ответа, суточный перерыв.
final class UpdateCheckTests: XCTestCase {
    func testVersionParsing() {
        XCTAssertEqual(UpdateCheck.parse("0.7.0"), [0, 7, 0])
        XCTAssertEqual(UpdateCheck.parse("v1.2.3"), [1, 2, 3])
        XCTAssertEqual(UpdateCheck.parse(""), [0])
    }

    func testPrereleaseIsNotNewerThanRelease() {
        // «0.8.0-rc1» не должен выглядеть новее готовой 0.8.0.
        XCTAssertFalse(UpdateCheck.isNewer("0.8.0-rc1", than: "0.8.0"))
        XCTAssertTrue(UpdateCheck.isNewer("0.8.0", than: "0.7.9"))
        XCTAssertFalse(UpdateCheck.isNewer("0.7.0", than: "0.7.0"))
        XCTAssertTrue(UpdateCheck.isNewer("1.0", than: "0.9.9"))
        XCTAssertFalse(UpdateCheck.isNewer("0.9", than: "0.9.1"))
    }

    func testReleaseIsReadFromTheAnswer() {
        let json = """
        {"tag_name": "v0.9.0", "html_url": "https://example/0.9.0", "body": " что нового "}
        """.data(using: .utf8)!
        let release = UpdateCheck.release(fromJSON: json)
        XCTAssertEqual(release?.version, "0.9.0")
        XCTAssertEqual(release?.url?.absoluteString, "https://example/0.9.0")
        XCTAssertEqual(release?.notes, "что нового")
    }

    func testAnswerWithoutATagIsIgnored() {
        let json = #"{"message": "Not Found"}"#.data(using: .utf8)!
        XCTAssertNil(UpdateCheck.release(fromJSON: json))
        XCTAssertNil(UpdateCheck.release(fromJSON: Data("не json".utf8)))
    }

    func testIntervalIsRespected() {
        let defaults = UserDefaults(suiteName: "glyphstroke.tests.interval")!
        defaults.removePersistentDomain(forName: "glyphstroke.tests.interval")

        XCTAssertTrue(UpdateCheck.due(defaults: defaults), "ни разу не спрашивали")
        UpdateCheck.remember(nil, defaults: defaults)
        XCTAssertFalse(UpdateCheck.due(defaults: defaults))
        let later = Date().addingTimeInterval(UpdateCheck.interval + 1)
        XCTAssertTrue(UpdateCheck.due(now: later, defaults: defaults))
        // время перевели назад: лучше спросить, чем ждать сутки от будущего
        let earlier = Date().addingTimeInterval(-10_000)
        XCTAssertTrue(UpdateCheck.due(now: earlier, defaults: defaults))
    }

    func testPendingKeepsOnlyNewerVersions() {
        let defaults = UserDefaults(suiteName: "glyphstroke.tests.pending")!
        defaults.removePersistentDomain(forName: "glyphstroke.tests.pending")

        let release = UpdateCheck.Release(version: "9.9.9",
                                          url: URL(string: "https://example/9.9.9"),
                                          notes: "")
        UpdateCheck.remember(release, defaults: defaults)
        XCTAssertEqual(UpdateCheck.pending(currentVersion: "0.2.0", defaults: defaults)?.version,
                       "9.9.9")
        XCTAssertNil(UpdateCheck.pending(currentVersion: "9.9.9", defaults: defaults),
                     "своя же версия — не новость")
        XCTAssertNil(UpdateCheck.pending(currentVersion: "10.0.0", defaults: defaults))
    }

    func testAddressesAreBuiltFromTheRepository() {
        XCTAssertEqual(UpdateCheck.apiURL(repository: "runetrace/glyphstroke")?.absoluteString,
                       "https://api.github.com/repos/runetrace/glyphstroke/releases/latest")
        XCTAssertEqual(UpdateCheck.releasePage(repository: "a/b")?.absoluteString,
                       "https://github.com/a/b/releases/latest")
    }
}
